server/src/test/java/org/opensearch/index/engine/NRTReplicationEngineTests.java

```diff
@@ -367,4 +370,63 @@ private NRTReplicationEngine buildNrtReplicaEngine(AtomicLong globalCheckpoint,
     private NRTReplicationEngine buildNrtReplicaEngine(AtomicLong globalCheckpoint, Store store) throws IOException {
         return buildNrtReplicaEngine(globalCheckpoint, store, defaultSettings);
     }
+
+    public void testGetSegmentInfosSnapshotPreservesFilesUntilRelease() throws Exception {
+        final AtomicLong globalCheckpoint = new AtomicLong(SequenceNumbers.NO_OPS_PERFORMED);
+
+        // TODO: Remove this divergent commit logic and copy Segments_N from primary with node-node.
+        // randomly toggle commit / no commit.
+        IndexSettings settings = REMOTE_STORE_INDEX_SETTINGS;
+        final boolean shouldCommit = randomBoolean();
+        if (shouldCommit) {
+            settings = INDEX_SETTINGS;
+        }
+        try (
+            final Store nrtEngineStore = createStore(REMOTE_STORE_INDEX_SETTINGS, newDirectory());
+            final NRTReplicationEngine nrtEngine = buildNrtReplicaEngine(globalCheckpoint, nrtEngineStore, settings)
+        ) {
+            List<Engine.Operation> operations = generateHistoryOnReplica(between(5, 10), randomBoolean(), randomBoolean(), randomBoolean());
+            for (Engine.Operation op : operations) {
+                applyOperation(engine, op);
+                applyOperation(nrtEngine, op);
+                // refresh to create a lot of segments.
+                engine.refresh("test");
+            }
+            // wipe the nrt directory initially so we can sync with primary.
+            Lucene.cleanLuceneIndex(nrtEngineStore.directory());
```

Comment:
nit: can we assert there are no files post clean up

Reply:
I think this would be a bit redundant, the copyFrom on the next step would throw a FileAlreadyExists if it clean didn't do its job.